src/docfx/lib/MarkdownUtility.cs

```diff
@@ -33,53 +32,79 @@ internal static class MarkdownUtility
+                .UseYamlFrontMatter()
+                .UseDocfxExtensions(markdownContext)
+                .UseExtractYamlHeader(file, errors, metadata)
+                .UseExtractTitle(title)
+                .UseResolveHtmlLinks(markdownContext, hasHtml)
+                .Build();
+
             using (InclusionContext.PushFile(file))
             {
-                var metadata = new StrongBox<JObject>();
-                var title = new StrongBox<string>();
-                var hasHtml = new StrongBox<bool>();
-
-                var markdownContext = new MarkdownContext(
-                    s_markdownTokens,
-                    context.ReportWarning,
-                    context.ReportError,
-                    ReadFile,
-                    GetLink);
-
-                var pipeline = new MarkdownPipelineBuilder()
-                    .UseYamlFrontMatter()
-                    .UseDocfxExtensions(markdownContext)
-                    .UseExtractYamlHeader(context, file, metadata)
-                    .UseExtractTitle(title)
-                    .UseResolveHtmlLinks(markdownContext, hasHtml)
-                    .Build();
-
                 var html = Markdown.ToHtml(markdown, pipeline);
 
                 var result = new MarkupResult
                 {
                     Title = title.Value,
                     HasHtml = hasHtml.Value,
                     Metadata = metadata.Value,
+                    Errors = errors,
                 };
 
                 return (html, result);
             }
 
+            void LogError(string code, string message, string doc, int line)
+            {
+                errors.Add(new DocfxException(ReportLevel.Error, code, message, doc, line));
+            }
+
+            void LogWarning(string code, string message, string doc, int line)
+            {
+                errors.Add(new DocfxException(ReportLevel.Warning, code, message, doc, line));
+            }
+
             (string content, object file) ReadFile(string path, object relativeTo)
             {
                 Debug.Assert(relativeTo is Document);
 
-                return ((Document)relativeTo).TryResolveContent(path);
+                var (error, content, child) = ((Document)relativeTo).TryResolveContent(path);
+
+                if (error != null)
+                {
+                    errors.Add(error);
+                }
+
+                return (content, child);
             }
 
             string GetLink(string path, object relativeTo)
             {
                 Debug.Assert(relativeTo is Document);
 
-                return resolveHref((Document)relativeTo, path, file);
+                var (error, link, child) = ((Document)relativeTo).TryResolveHref(path, file);
+
+                if (error != null)
+                {
+                    errors.Add(error);
+                }
+
+                if (child != null)
+                {
+                    buildChild(child);
```

Comment:
>buildChild [](start = 20, length = 10)

i still dones't agree to put the build context down to markdown lib, would we move the markdown utility build/markdown?

Reply:
it will be moved to build, after you checkin dep map

---
In reply to: [193959523](https://github.com/dotnet/docfx/pull/2857#discussion_r193959523) [](ancestors = 193959523)